contracts/hydro/src/utils.rs

```diff
@@ -0,0 +1,290 @@
+        }
+    })
+}
+
+// This function validates if user should be allowed to lock more tokens, depending on the total amount of
+// currently locked tokens, existence of the extra cap and the number of tokens user already locked in that cap.
+// Caps that we consider:
+//      1) Total cap: the maximum number of tokens that can be locked in the contract, regardless of the
+//         extra cap existence.
+//      2) Extra cap: the number of tokens reserved for users that had some voting power in previous round.
+//         It will be available to those users to lock additional tokens even if the public cap is filled.
+//         The extra cap will be available only for certain time frame at the beginning of the round. Duration
+//         will be determined through governance. After the extra cap duration has expired, any tokens left up
+//         to the total cap are allowed to be locked by any Hydro user.
+//      3) Public cap: during the existence of extra cap, new Hydro users will be allowed to lock tokens only
+//         in public cap, where public_cap = total_cap - extra_cap.
+//         After the extra cap duration expires, public cap becomes equal to the total cap.
+pub fn validate_locked_tokens_caps(
+    deps: &DepsMut<NeutronQuery>,
+    constants: &Constants,
+    current_round: u64,
+    sender: &Addr,
+    total_locked_tokens: u128,
+    amount_to_lock: u128,
+) -> Result<LockingInfo, ContractError> {
+    let lock_limit_reached_error = Err(ContractError::Std(StdError::generic_err(
+        "The limit for locking tokens has been reached. No more tokens can be locked.",
+    )));
+
+    // Regardless of public_cap and extra_cap, we must make sure that nobody can lock more than total_cap
+    let total_locked_after = total_locked_tokens + amount_to_lock;
+    if total_locked_after > constants.max_locked_tokens {
+        return lock_limit_reached_error;
+    }
+
+    let public_cap = constants.max_locked_tokens - constants.current_users_extra_cap;
+
+    // This branch will be executed in one of the following cases:
+    //   1) constants.current_users_extra_cap != 0 and there is SOME room in the public cap
+    //   2) constants.current_users_extra_cap == 0 and there is ENOUGH room in the public_cap.
+    //      Since in this case public_cap == total_cap, there must be enough room in the public_cap,
+    //      otherwise we would error out at the start of this function.
+    if public_cap > total_locked_tokens {
+        // Check if entire amount_to_lock can fit into a public cap
+        if public_cap >= total_locked_after {
+            return Ok(LockingInfo {
+                lock_in_public_cap: Some(amount_to_lock),
+                lock_in_known_users_cap: None,
+            });
+        }
+
+        // Lock as much as possible in the public_cap, and the rest in the extra_cap
+        let lock_in_public_cap = public_cap - total_locked_tokens;
+        let lock_in_known_users_cap = amount_to_lock - lock_in_public_cap;
```

Comment:
Suggest to make the naming consistent and call it either known_users_cap or _extra_cap everywhere.
probably, calling it known_users_cap seems better, actually!